pkg/skaffold/build/scheduler.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package build
+
+import (
+	"context"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+)
+
+// scheduler handles build ordering and concurrency.
+type scheduler struct {
+	dagMap map[string]artifactDAG
+	// sem is a buffered channel that will allow up to `c` concurrent operations when its buffer size is `c`.
+	sem chan bool
+}
+
+func newScheduler(artifacts []*latest.Artifact, concurrency int) scheduler {
+	if concurrency == 0 {
```

Comment:
For current version, i thought we would always have non-parallel builds?

Reply:
we'll support concurrent builds as we do currently- the build logs are printed sequentially, though independent builds can run in parallel.

Reply:
to give more detail: we set the default concurrency to 1 in `defaults.go`. So concurrency `0` is the user's way of explicitly saying `max concurrency`. This is the existing code behavior and it works for the new scenario of build dependencies too. 
Currently in master, the output logs are displayed in the order specified in the input `latest.Artifact` slice. I think that was mostly a bug since if the concurrency is 1 and order of execution is artifact4, artifact2, artifact3 and artifact1 then we won't see any logs until artifact1 starts building and then they are all shown together.
I fix this in [another](https://github.com/GoogleContainerTools/skaffold/pull/4843) PR where the logs are printed in the actual build order. But I keep the current behavior of displaying concurrently run and concluded build logs together one after the other immediately. This behavior can later be substituted with a progress bar or multi drop-down writers to show all concurrent builds together but that is left as a future work item.
These are described in the design doc [here](https://github.com/GoogleContainerTools/skaffold/blob/master/docs/design_proposals/artifact-dependencies.md#build-logs-reporting)